### PR TITLE
[unitaryHACK] Added SWAP gate decomposition and related tests

### LINKED ISSUE #1304: [unitaryhack] Define decomposition for `qml.SWAP`
*This issue has been tagged for contributions during [unitaryHACK](https://unitaryfund.github.io/unitaryhack/participating-projects.html)*

Each PennyLane operation defines a decomposition. The decomposition of an operation is an ordered list of other operations. Applying this list of operations evolves quantum states the same way as applying the operation itself does. In most cases, this list includes elementary gates that are often supported by most quantum devices.

We would like to add the decomposition of the `qml.SWAP` operation by:
1. Defining the decomposition method for [the `qml.SWAP` operation](https://github.com/PennyLaneAI/pennylane/blob/1930230dde60112ad90af0951fa9295d374e1adf/pennylane/ops/qubit.py#L528) that has the following signature:
```python
    @staticmethod
    def decomposition(wires):
```
Other operations in the `pennylane/ops/qubit.py` file can serve as an example for defining this method (e.g., [see `qml.Hadamard`](https://github.com/PennyLaneAI/pennylane/blob/1930230dde60112ad90af0951fa9295d374e1adf/pennylane/ops/qubit.py#L87)).
2. Including a test case for the decomposition of `qml.SWAP` in the `tests/ops/test_qubit_ops.py` file to verify its correctness.

### PR DESCRIPTION
**Pull request by charmerDark**
**Context:**
Added SWAP decomposition and wrote tests to it

**Description of the Change:**
added the the function 
 ``` 
  @staticmethod
    def decomposition(wires): 
```
to ` pennylane/ops/qubit.py `  and tests for it at ` tests/ops/test_qubit_ops.py `

**Possible Drawbacks:**
similar to #1313 
**Related GitHub Issues:**
resolves #1304

### CODE REVIEW

File: .github/CHANGELOG.md [hunk 1]
```diff
@@ -250,7 +252,7 @@ random_mat2 = rng.standard_normal(3, requires_grad=False)
 
 This release contains contributions from (in alphabetical order):
 
-Thomas Bromley, Olivia Di Matteo, Diego Guala, Anthony Hayes, Josh Izaac, Brian Shi, Antal Száva, Pavan Jayasinha
+Thomas Bromley, Olivia Di Matteo, Diego Guala, Anthony Hayes, Josh Izaac, Brian Shi, Antal Száva, Pavan Jayasinha, Vishnu Ajith.
```
Comment (antalszava) [inline .github/CHANGELOG.md:255]: just to aim for the alphabetical order
```suggestion
Vishnu Ajith, Thomas Bromley, Olivia Di Matteo, Diego Guala, Anthony Hayes, Josh Izaac, Brian Shi, Antal Száva, Pavan Jayasinha
```

File: tests/ops/test_qubit_ops.py [hunk 2]
```diff
@@ -923,6 +923,19 @@ def test_iswap_eigenval(self):
...
         assert np.allclose(res, exp)
 
+    def test_swap_decomposition(self):
+        "Tests the swap operator produces the correct output"
+        opr = qml.SWAP(wires=[0, 1])
+        decomp = opr.decomposition([0, 1])
+        mat = []
+        for op in reversed(decomp):
+            if isinstance(op, qml.CNOT) and op.wires.tolist() == [0, 1]:
+                mat.append(np.array([[1, 0, 0, 0], [0, 1, 0, 0], [0, 0, 0, 1], [0, 0, 1, 0]]))
```
Comment (antalszava) [inline tests/ops/test_qubit_ops.py:933]: We could use the `CNOT` matrix import from `gate_data` at the top:
```python
from gate_data import (
    (...)
    CNOT,
```

```suggestion
                mat.append(CNOT)
```

File: tests/ops/test_qubit_ops.py [hunk 3]
```diff
@@ -923,6 +923,19 @@ def test_iswap_eigenval(self):
         res = op.eigvals
         assert np.allclose(res, exp)
 
+    def test_swap_decomposition(self):
+        "Tests the swap operator produces the correct output"
```
Comment (antalszava) [inline tests/ops/test_qubit_ops.py:927]:
```suggestion
        """Tests the swap operator produces the correct output"""
```

**Review by antalszava — APPROVED** [review]
Hi @charmerDark, this looks good to me! :tada: :slightly_smiling_face:  Extra credit for going through the challenges with `git`. Thank you for your contribution! :)

File: tests/ops/test_qubit_ops.py [hunk 4]
```diff
@@ -923,6 +923,19 @@ def test_iswap_eigenval(self):
         res = op.eigvals
         assert np.allclose(res, exp)
 
+    def test_swap_decomposition(self):
+        """Tests the swap operator produces the correct output"""
+        opr = qml.SWAP(wires=[0, 1])
+        decomp = opr.decomposition([0, 1])
```
Comment (antalszava) [inline tests/ops/test_qubit_ops.py:929]:
```suggestion
        decomp = opr.decomposition([0, 1])

```

File: tests/ops/test_qubit_ops.py [hunk 5]
```diff
@@ -923,6 +923,19 @@ def test_iswap_eigenval(self):
...
+        mat = []
+        for op in reversed(decomp):
+            if isinstance(op, qml.CNOT) and op.wires.tolist() == [0, 1]:
+                mat.append(CNOT)
+            elif isinstance(op, qml.CNOT) and op.wires.tolist() == [1, 0]:
+                mat.append(np.array([[1, 0, 0, 0], [0, 0, 0, 1], [0, 0, 1, 0], [0, 1, 0, 0]]))
+        decomposed_matrix = np.linalg.multi_dot(mat)
```
Comment (antalszava) [inline tests/ops/test_qubit_ops.py:937]:
```suggestion
        decomposed_matrix = np.linalg.multi_dot(mat)

```

File: tests/ops/test_qubit_ops.py [hunk 6]
```diff
@@ -923,6 +923,20 @@ def test_iswap_eigenval(self):
...
+        """Tests the swap operator produces the correct output"""
+        opr = qml.SWAP(wires=[0, 1])
+        decomp = opr.decomposition([0, 1])
+
+        mat = []
+        for op in reversed(decomp):
+            if isinstance(op, qml.CNOT) and op.wires.tolist() == [0, 1]:
+                mat.append(CNOT)
+            elif isinstance(op, qml.CNOT) and op.wires.tolist() == [1, 0]:
+                mat.append(np.array([[1, 0, 0, 0], [0, 0, 0, 1], [0, 0, 1, 0], [0, 1, 0, 0]]))
```
Comment (antalszava) [inline tests/ops/test_qubit_ops.py:936]:
```suggestion
                mat.append(np.array([[1, 0, 0, 0], [0, 0, 0, 1], [0, 0, 1, 0], [0, 1, 0, 0]]))

```

**Comment by charmerDark** [comment]
Thank you so much @antalszava Couldn't have made it without all your support. I learnt so much from this. Hope to be able to contribute more to pennylane soon.

**Pull request merged**